File: bindings/kotlin/whatwg/src/main/kotlin/com/whatwg/providers/NetworkProvider.kt
package com.whatwg.providers

/**
 * Interface for providing network/fetch functionality.
 *
 * Implement this interface to provide HTTP request support.
 */
interface NetworkProvider {

    /**
     * Performs an HTTP request.
     *
     * @param request The request configuration.
     * @return The response.
     * @throws Exception If the request fails.
     */
    suspend fun fetch(request: NetworkRequest): NetworkResponse

    /**
     * Checks if the device is online.
     *
     * @return `true` if network is available.
     */
    fun isOnline(): Boolean
}

/**
 * Configuration for a network request.
 */
data class NetworkRequest(
    /**
     * The request URL.
     */
    val url: String,
    
    /**
     * The HTTP method.
     */
    val method: String = "GET",
    
    /**
     * Request headers.
     */
    val headers: Map<String, String> = emptyMap(),
    
    /**
     * Request body.
     */
    val body: ByteArray? = null,
    
    /**
     * Request timeout in milliseconds.
     */
    val timeoutMs: Long = 30000
) {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false
        other as NetworkRequest
        return url == other.url &&
                method == other.method &&
                headers == other.headers &&
                body.contentEquals(other.body) &&
                timeoutMs == other.timeoutMs
    }

    override fun hashCode(): Int {
        var result = url.hashCode()
        result = 31 * result + method.hashCode()
        result = 31 * result + headers.hashCode()
        result = 31 * result + (body?.contentHashCode() ?: 0)
        result = 31 * result + timeoutMs.hashCode()
        return result
    }
}

/**
 * Response from a network request.
 */
data class NetworkResponse(
    /**
     * HTTP status code.
     */
    val statusCode: Int,
    
    /**
     * Response headers.
     */
    val headers: Map<String, String>,
    
    /**
     * Response body.
     */
    val body: ByteArray
) {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false
        other as NetworkResponse
        return statusCode == other.statusCode &&
                headers == other.headers &&
                body.contentEquals(other.body)
    }

    override fun hashCode(): Int {
        var result = statusCode
        result = 31 * result + headers.hashCode()
        result = 31 * result + body.contentHashCode()
        return result
    }
}

File: bindings/swift/Sources/WhatWG/Providers/NetworkProvider.swift
import Foundation

/// Protocol for providing network/fetch functionality.
///
/// Implement this protocol to provide HTTP request support.
/// The default implementation uses URLSession.
///
public protocol NetworkProvider: AnyObject, Sendable {
    
    /// Performs an HTTP request.
    ///
    /// - Parameter request: The request configuration.
    /// - Returns: The response.
    /// - Throws: If the request fails.
    func fetch(request: NetworkRequest) async throws -> NetworkResponse
    
    /// Checks if the device is online.
    ///
    /// - Returns: `true` if network is available.
    func isOnline() -> Bool
}

/// Configuration for a network request.
public struct NetworkRequest: Sendable {
    /// The request URL.
    public var url: URL
    
    /// The HTTP method.
    public var method: String
    
    /// Request headers.
    public var headers: [String: String]
    
    /// Request body.
    public var body: Data?
    
    /// Request timeout in seconds.
    public var timeout: TimeInterval
    
    /// Creates a new network request.
    public init(
        url: URL,
        method: String = "GET",
        headers: [String: String] = [:],
        body: Data? = nil,
        timeout: TimeInterval = 30
    ) {
        self.url = url
        self.method = method
        self.headers = headers
        self.body = body
        self.timeout = timeout
    }
}

/// Response from a network request.
public struct NetworkResponse: Sendable {
    /// HTTP status code.
    public var statusCode: Int
    
    /// Response headers.
    public var headers: [String: String]
    
    /// Response body.
    public var body: Data
    
    /// Creates a new network response.
    public init(statusCode: Int, headers: [String: String], body: Data) {
        self.statusCode = statusCode
        self.headers = headers
        self.body = body
    }
}
